ledger/src/blockstore.rs

```diff
@@ -5390,6 +5401,62 @@ impl Blockstore {
         Ok(())
     }
 
+    /// Handles chaining updates when an UpdateParent marker overrides a previously
+    /// set parent from a BlockHeader.
+    ///
+    /// TODO(ashwin, ksn): this chaining only occurs for `SlotMeta`s in the column associated with
+    /// `BlockLocation::Original`
+    fn update_chaining_for_parent_metas(
+        &self,
+        working_set: &mut HashMap<(BlockLocation, u64), SlotMetaWorkingSetEntry>,
+        parent_metas: &HashMap<(BlockLocation, u64), ParentMetaWorkingSetEntry>,
+        new_chained_slots: &mut HashMap<u64, Rc<RefCell<SlotMeta>>>,
+    ) -> Result<()> {
+        // Filter to dirty UpdateParent entries for Original location only
+        let update_parent_entries = parent_metas.iter().filter_map(|((location, slot), wse)| {
+            if !matches!(location, BlockLocation::Original) {
+                return None;
+            }
+
+            match &wse.parent_meta {
+                Some(WorkingEntry::Dirty(pm)) if pm.populated_from_update_parent() => {
+                    Some((*slot, pm.parent_slot))
+                }
+                _ => None,
+            }
+        });
+
+        for (slot, new_parent_slot) in update_parent_entries {
+            // Get old parent from slot meta; update parent_slot to new value
+            let old_parent_slot =
+                if let Some(entry) = working_set.get_mut(&(BlockLocation::Original, slot)) {
+                    let old = entry.old_slot_meta.as_ref().and_then(|m| m.parent_slot);
+                    entry.new_slot_meta.borrow_mut().parent_slot = Some(new_parent_slot);
+                    old
+                } else {
+                    None
+                };
+
+            // Remove slot from old parent's next_slots if parent changed
+            if let Some(old_parent) = old_parent_slot.filter(|&old| old != new_parent_slot) {
+                self.find_slot_meta_else_create(working_set, new_chained_slots, old_parent)?
+                    .borrow_mut()
+                    .next_slots
+                    .retain(|&s| s != slot);
+            }
+
+            // Add slot to new parent's next_slots
+            let new_parent_meta =
+                self.find_slot_meta_else_create(working_set, new_chained_slots, new_parent_slot)?;
+            let mut new_parent = new_parent_meta.borrow_mut();
+            if !new_parent.next_slots.contains(&slot) {
+                new_parent.next_slots.push(slot);
```

Comment:
forgot how painful this logic is haha

this new_parent might not be in the chained cache, so it won't be committed here https://github.com/ksn6/alpenglow/blob/59a53d6cf818f6cc774713668f12bef4925e4e49/ledger/src/blockstore.rs#L5314-L5316

But it will be committed later by the dirty SlotMeta working set?